fqn_decorators/asynchronous.py

```diff
@@ -9,19 +9,27 @@
 
 
 class AsyncDecorator(Decorator):
-    async def __call__(self, *args, **kwargs):
-        self.fqn = self.get_fqn()
-        self.args = args
-        self.kwargs = kwargs
+    def __new__(cls, func=None, **params):
+        def _wrap_func(self, f, *args, **kwargs):
+            return self.__class__(f, **self.params)
 
-        self.before()
-        try:
-            self.result = await self.func(*self.args, **self.kwargs)
-        except:
-            self.exc_info = sys.exc_info()
-            self.exception()
-            raise
-        finally:
-            self.after()
+        async def async_call(self, *args, **kwargs):
+            self.fqn = self.get_fqn()
+            self.args = args
+            self.kwargs = kwargs
+            self.before()
+            try:
+                self.result = await self.func(*self.args, **self.kwargs)
+            except:
+                self.exc_info = sys.exc_info()
+                self.exception()
+                raise
+            finally:
+                self.after()
+            return self.result
 
-        return self.result
+        # Create __call__ method dynamically to allow passing parameters to the decorator.
+        # Otherwise the decorated method is treated as a coroutine, which can't be called.
+        setattr(AsyncDecorator, '__call__', async_call if func else _wrap_func)
```

Comment:
I think saving this on a class attribute does something other than you expect, unless you want to achieve the following behaviour:

```
class MyClass:
    def __new__(cls, test_string):
        setattr(MyClass, 'test', test_string)
        return super().__new__(cls)

a1 = MyClass("will this value be shown for a1?")
a2 = MyClass("and this for a2?")

print(a1.test)
print(a2.test)
```

which will yield the following output:
```
and this for a2?
and this for a2?
```

I think this is not your intention, is it?